library/std/src/os/windows/io/handle.rs

```diff
@@ -0,0 +1,390 @@
+//! Owned and borrowed OS handles.
+
+#![unstable(feature = "io_safety", issue = "87074")]
+
+use super::raw::{AsRawHandle, FromRawHandle, IntoRawHandle, RawHandle};
+use crate::convert::TryFrom;
+use crate::ffi::c_void;
+use crate::fmt;
+use crate::fs;
+use crate::marker::PhantomData;
+use crate::mem::forget;
+use crate::ptr::NonNull;
+use crate::sys::c;
+use crate::sys_common::{AsInner, FromInner, IntoInner};
+
+/// A borrowed handle.
+///
+/// This has a lifetime parameter to tie it to the lifetime of something that
+/// owns the handle.
+///
+/// This uses `repr(transparent)` and has the representation of a host handle,
+/// so it can be used in FFI in places where a handle is passed as an argument,
+/// it is not captured or consumed, and it is never null.
+///
+/// Note that it *may* have the value `INVALID_HANDLE_VALUE`. See [here] for
```

Comment:
Maybe saying something along the lines of "The handle must always be valid. Note that in some cases `INVALID_HANDLE_VALUE` is considered a valid value, see [blog post]. The creator of `BorrowedHandle` must ensure the handle is valid".

Because if I read it correctly it just happens that `INVALID_HANDLE_VALUE` overlaps with some valid handles? So when the handle is returned by the OS it should be checked by the caller with call-specific knowledge, i.e. whether or not `INVALID_HANDLE_VALUE` is considered valid.

Reply:
I've now added comments along these lines.